File: client/archive/android/app/src/main/java/com/solutioncrafts/BriskVTU/feature/wallet/WalletApi.kt
package com.solutioncrafts.BriskVTU.feature.wallet

import kotlinx.serialization.Serializable
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path

@Serializable
data class WalletResponse(
    val mainBalanceNgn: Double,
    val cashbackBalanceNgn: Double
)

@Serializable
data class WalletConfigResponse(
    val exchangeRate: Double,
    val topUpFeePercentage: Double
)

@Serializable
data class PaymentCardResponse(
    val id: String,
    val brand: String,
    val last4: String,
    val expMonth: Int,
    val expYear: Int,
    val isDefault: Boolean
)

@Serializable
data class SetupIntentResponse(
    val clientSecret: String
)

@Serializable
data class TopUpResponse(
    val transactionId: String,
    val clientSecret: String,
    val amountUsd: Double,
    val serviceFeeNgn: Double,
    val exchangeRate: Double
)

@Serializable
data class TopUpRequest(
    val amountUsd: Double,
    val cardId: String? = null
)

@Serializable
data class AttachCardRequest(
    val stripePaymentMethodId: String
)

interface WalletApi {
    @GET("wallet")
    suspend fun getWallet(): WalletResponse

    @GET("wallet/config")
    suspend fun getConfig(): WalletConfigResponse

    @POST("wallet/initiate")
    suspend fun initiateWallet(): WalletResponse

    @GET("wallet/cards")
    suspend fun listCards(): List<PaymentCardResponse>

    @POST("wallet/cards/setup-intent")
    suspend fun createSetupIntent(): SetupIntentResponse

    @POST("wallet/cards")
    suspend fun attachCard(@Body request: AttachCardRequest): PaymentCardResponse

    @DELETE("wallet/cards/{cardId}")
    suspend fun removeCard(@Path("cardId") cardId: String)

    @POST("wallet/topup/initiate")
    suspend fun initiateTopUp(@Body request: TopUpRequest): TopUpResponse
}

File: client/archive/ios/BriskVTU/Features/Wallet/WalletAPI.swift
import Foundation

enum WalletAPIError: Error {
    case invalidResponse
    case decodingFailed
}

struct WalletResponse: Codable {
    let mainBalanceNgn: Double
    let cashbackBalanceNgn: Double
}

struct WalletConfigResponse: Codable {
    let exchangeRate: Double
    let topUpFeePercentage: Double
}

struct PaymentCardResponse: Codable, Identifiable {
    let id: String
    let brand: String
    let last4: String
    let expMonth: Int
    let expYear: Int
    let isDefault: Bool
}

struct SetupIntentResponse: Codable {
    let clientSecret: String
}

struct TopUpResponse: Codable {
    let transactionId: String
    let clientSecret: String
    let amountUsd: Double
    let serviceFeeNgn: Double
    let exchangeRate: Double
}

struct TopUpRequest: Codable {
    let amountUsd: Double
    let cardId: String?
}

struct AttachCardRequest: Codable {
    let stripePaymentMethodId: String
}

final class WalletAPI {
    static let shared = WalletAPI()
    
    private let client = APIClient.shared
    
    func getWallet() async throws -> WalletResponse {
        let res: BaseResponse<WalletResponse> = try await client.request(endpoint: "/wallet", method: "GET")
        guard let wallet = res.data else { throw APIError.noData }
        return wallet
    }
    
    func getConfig() async throws -> WalletConfigResponse {
        let res: BaseResponse<WalletConfigResponse> = try await client.request(endpoint: "/wallet/config", method: "GET")
        guard let config = res.data else { throw APIError.noData }
        return config
    }
    
    func initiateWallet() async throws -> WalletResponse {
        let res: BaseResponse<WalletResponse> = try await client.request(endpoint: "/wallet/initiate", method: "POST")
        guard let wallet = res.data else { throw APIError.noData }
        return wallet
    }
    
    func listCards() async throws -> [PaymentCardResponse] {
        let res: BaseResponse<[PaymentCardResponse]> = try await client.request(endpoint: "/wallet/cards", method: "GET")
        return res.data ?? []
    }
    
    func createSetupIntent() async throws -> SetupIntentResponse {
        let res: BaseResponse<SetupIntentResponse> = try await client.request(endpoint: "/wallet/cards/setup-intent", method: "POST")
        guard let si = res.data else { throw APIError.noData }
        return si
    }
    
    func attachCard(paymentMethodId: String) async throws -> PaymentCardResponse {
        let req = AttachCardRequest(stripePaymentMethodId: paymentMethodId)
        let data = try JSONEncoder().encode(req)
        let res: BaseResponse<PaymentCardResponse> = try await client.request(endpoint: "/wallet/cards", method: "POST", body: data)
        guard let card = res.data else { throw APIError.noData }
        return card
    }
    
    func removeCard(cardId: String) async throws {
        let _: BaseResponse<EmptyResponse> = try await client.request(endpoint: "/wallet/cards/\(cardId)", method: "DELETE")
    }
    
    func initiateTopUp(amountUsd: Double, cardId: String?) async throws -> TopUpResponse {
        let req = TopUpRequest(amountUsd: amountUsd, cardId: cardId)
        let data = try JSONEncoder().encode(req)
        let res: BaseResponse<TopUpResponse> = try await client.request(endpoint: "/wallet/topup/initiate", method: "POST", body: data)
        guard let topUp = res.data else { throw APIError.noData }
        return topUp
    }
}
